cmd/opentelemetry/app/exporter/elasticsearchexporter/spanstore.go

```diff
@@ -216,11 +223,36 @@ func hashCode(serviceName, operationName string) string {
 
 type bulkItem struct {
 	// span associated with the bulk operation
-	span *dbmodel.Span
+	spanData esmodeltranslator.ConvertedData
 	// isService indicates that this bulk operation is for service index
 	isService bool
 }
 
 func (w *esSpanWriter) esClientVersion() int {
 	return w.client.MajorVersion()
 }
+
+func bulkItemsToTraces(bulkItems []bulkItem) pdata.Traces {
+	traces := pdata.NewTraces()
+	for _, op := range bulkItems {
+		spanData := op.spanData
+		if traces.ResourceSpans().Len() == 0 {
+			traces.ResourceSpans().Resize(1)
+		} else {
+			traces.ResourceSpans().Resize(traces.ResourceSpans().Len())
+		}
+		rss := traces.ResourceSpans().At(traces.ResourceSpans().Len() - 1)
+		if !spanData.Resource.IsNil() {
+			spanData.Resource.CopyTo(rss.Resource())
+		}
+		rss.InstrumentationLibrarySpans().Resize(1)
```

Comment:
Similar to above, you could keep some kind of map of Resource/InstrumentationLibrary => InstrumentationLibrarySpans which you'd then check to see if you had an InstrumentationLibrarySpans to append to or if you should just make a new one.

Might be faster to just make one ResourceSpans/InstrumentationLibrarySpans for each bulkitem and forget trying to reuse them.  It's hard to say if spans in a batch will likely come from the same trace and this will be worthwhile or not.   Probably depends on which processors are configured in the pipeline.

Reply:
We cannot expect that spans in a batch share the same resource&instrumentation.

> Might be faster to just make one ResourceSpans/InstrumentationLibrarySpans for each bulkitem and forget trying to reuse them. It's hard to say if spans in a batch will likely come from the same trace and this will be worthwhile or not. Probably depends on which processors are configured in the pipeline.

Keeping the resource/instrumentation -> spans would be more memory efficient. However, I would ho with this simple approach for now. The instrumentation library has two string fields and resource has a map.

Reply:
Agreed +1